prow/git/v2/client_factory.go

```diff
@@ -289,7 +293,7 @@ func (c *clientFactory) ClientFor(org, repo string) (RepoClient, error) {
 		return nil, err
 	}
 
-	repoDir, err := os.MkdirTemp(c.cacheDirBase, "gitrepo")
+	repoDir, err := os.MkdirTemp(*defaultTempDir(), "gitrepo")
```

Comment:
I don't understand what's going on. On one hand the switch statement is being refactored out into its own function.  On the other we're calling this switch statement twice now instead of just once. How is it different from the previous behavior? Can you try ELI5 for those who are not familiar with this area of the codebase?

If you're able to come up with an explanation, can you also put that in the commit description?

Reply:
To clarify, I think it would help if you can explain /why/ this change is needed (not so much the /how/ part which I think I might have been emphasizing in my previous comment).

Reply:
Added some comments. In the case where `cacheDirBase == ""` the cacheDirBase gets defaulted to a temp dir. This code refactored that defaulting logic out so we can also use it for the location where the repo copies are going to go. We want repo copies to go in a temp dir so they don't fill up the cache.

Reply:
I agree with @listx here. `c.cacheDirBase` is already defaulted. Can you please elaborate more on what you are trying to achieve? Also, it will be helpful if you include/change the unit tests to prove the desired behavior.

Reply:
When c.cacheDirBase is NOT the default we still want the repo to be made in the default temp dir.
cacheDirBase will not always be default. We want repos to always be put in the default temp dir

Reply:
> Added some comments. In the case where `cacheDirBase == ""` the cacheDirBase gets defaulted to a temp dir.

More precisely `nil` not `""` but yes, I agree that this is the old behavior.

> This code refactored that defaulting logic out so we can also use it for the location where the repo copies are going to go. We want repo copies to go in a temp dir so they don't fill up the cache.

Are you saying that the `cacheDirBase` should always be a subdirectory inside `/var/tmp` because that way, there's no risk of adding too many repo copies in here because the OS will periodically (automatically) garbage-collect whatever is in `/var/tmp`? If so, why not just prepend `/var/tmp` as a prefix to cacheDirBase if it isn't there already?

Separate nit: it would be nice to also paste in the PR description you've updated into the commit message (you can do this with `git amend`) so that developers using the raw `git` CLI won't have to open up the associated merge commit and PR on GitHub to see the full explanation. This is what I meant by commit description in my earlier comment (and not the PR description).

Reply:
No. CacheDirBase defaults to /var/temp, but can be somewhere else (for example in a pv). The repoDir here is a copy of what is inside the cache for use in creating a client. These should always be inside /var/tmp/ because if not they will fill up the pv (in the case we are using one).

Reply:
I'm still on shaky ground because I don't understand the exact relationship between CacheDirBase, repoDir, pv, and temp dirs (it would have been good to spell this out in the commit description). It sounds like the existing behavior is broken --- is this what you're implying? For example the `repoDir` here is no longer tied to cacheDirBase --- but this behavioral change is necessary because the old behavior is broken...?

Anyway, since we're no longer using cacheDirBase here, where else is it still used?

Also, how come we have copies of repos to begin with? Is it for performance reasons? Or something else? Maybe to guarantee atomicity of git operations?

Reply:
Updated commit description to show the change in structure. And IIUC we have the repos cloned in the cache and then copy them to a new repoDir when we create a client so we can have multiple clients for the same repo not interfere with each other